File: ios/VelaWallet/VelaWallet/Services/WalletAPIService.swift
import Foundation

/// Client for getvela.app API endpoints.
final class WalletAPIService {
    static let baseURL = "https://getvela.app/api"

    // MARK: - Wallet Balances

    /// Fetch token balances across all supported networks.
    func fetchTokens(address: String) async throws -> [APIToken] {
        var components = URLComponents(string: "\(Self.baseURL)/wallet")!
        components.queryItems = [URLQueryItem(name: "address", value: address)]

        // Bypass URLSession cache to always get fresh data
        var request = URLRequest(url: components.url!)
        request.cachePolicy = .reloadIgnoringLocalCacheData

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            debugLog("[API] /wallet failed: HTTP \((response as? HTTPURLResponse)?.statusCode ?? 0)")
            throw APIError.fetchFailed
        }

        let result = try JSONDecoder().decode(WalletResponse.self, from: data)
        let filtered = result.tokens.filter { !$0.spam }
        let total = filtered.reduce(0) { $0 + $1.usdValue }
        debugLog("[API] /wallet: \(filtered.count) tokens, total $\(String(format: "%.2f", total))")
        return filtered
    }

    // MARK: - Exchange Rate

    /// Fetch USD to target currency exchange rate.
    func fetchExchangeRate(currency: String = "CNY") async throws -> Double {
        var components = URLComponents(string: "\(Self.baseURL)/exchange-rate")!
        components.queryItems = [URLQueryItem(name: "currency", value: currency)]

        let (data, response) = try await URLSession.shared.data(from: components.url!)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            throw APIError.fetchFailed
        }

        let result = try JSONDecoder().decode(ExchangeRateResponse.self, from: data)
        return result.rate
    }

    // MARK: - NFTs

    /// Fetch NFTs across all supported networks.
    func fetchNFTs(address: String) async throws -> [APINFT] {
        var components = URLComponents(string: "\(Self.baseURL)/nft")!
        components.queryItems = [URLQueryItem(name: "address", value: address)]

        var request = URLRequest(url: components.url!)
        request.cachePolicy = .reloadIgnoringLocalCacheData

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            debugLog("[API] /nft failed: HTTP \((response as? HTTPURLResponse)?.statusCode ?? 0)")
            throw APIError.fetchFailed
        }

        let result = try JSONDecoder().decode(NFTResponse.self, from: data)
        debugLog("[API] /nft: \(result.nfts.count) NFTs")
        return result.nfts
    }

    // MARK: - Types

    struct WalletResponse: Decodable {
        let tokens: [APIToken]
    }

    struct NFTResponse: Decodable {
        let nfts: [APINFT]
    }

    struct ExchangeRateResponse: Decodable {
        let currency: String
        let rate: Double
    }

    enum APIError: Error, LocalizedError {
        case fetchFailed
        var errorDescription: String? { "Failed to fetch data from server." }
    }
}

// MARK: - API Token Model

struct APIToken: Decodable, Identifiable {
    let network: String
    let chainName: String
    let symbol: String
    let balance: String
    let decimals: Int
    let logo: String?
    let name: String
    let tokenAddress: String?
    let priceUsd: Double?
    let spam: Bool

    var id: String { "\(network)_\(tokenAddress ?? "native")_\(symbol)" }

    /// Is this a native token (ETH, BNB, AVAX, etc.)?
    var isNative: Bool { tokenAddress == nil }

    /// Balance as Double
    var balanceDouble: Double { Double(balance) ?? 0 }

    /// USD value of holdings
    var usdValue: Double {
        guard let price = priceUsd else { return 0 }
        return balanceDouble * price
    }

    /// Chain ID derived from network identifier
    var chainId: Int {
        switch network {
        case "eth-mainnet": 1
        case "arb-mainnet": 42161
        case "base-mainnet": 8453
        case "opt-mainnet": 10
        case "matic-mainnet": 137
        case "bnb-mainnet": 56
        case "avax-mainnet": 43114
        default: 1
        }
    }

    /// Logo URL for this token
    var logoURL: URL? {
        // If API provides a logo, use it
        if let logo, !logo.isEmpty, let url = URL(string: logo) {
            return url
        }
        // Native token: use chain logo
        if isNative {
            return URL(string: "https://ethereum-data.awesometools.dev/chainlogos/eip155-\(chainId).png")
        }
        // ERC20: use token logo from chain data
        if let addr = tokenAddress {
            return URL(string: "https://ethereum-data.awesometools.dev/assets/eip155-\(chainId)/\(addr)/logo.png")
        }
        return nil
    }
}

// MARK: - NFT Model

struct APINFT: Decodable, Identifiable {
    let network: String
    let chainName: String
    let contractAddress: String
    let tokenId: String
    let name: String?
    let description: String?
    let image: String?
    let tokenType: String
    let collectionName: String?
    let collectionImage: String?

    var id: String { "\(network)_\(contractAddress)_\(tokenId)" }

    var displayName: String { name ?? "\(collectionName ?? "NFT") #\(tokenId)" }

    var imageURL: URL? {
        guard let image, !image.isEmpty else { return nil }
        // Handle IPFS URLs
        if image.hasPrefix("ipfs://") {
            return URL(string: "https://ipfs.io/ipfs/" + image.dropFirst(7))
        }
        return URL(string: image)
    }

    var collectionImageURL: URL? {
        guard let collectionImage, !collectionImage.isEmpty else { return nil }
        if collectionImage.hasPrefix("ipfs://") {
            return URL(string: "https://ipfs.io/ipfs/" + collectionImage.dropFirst(7))
        }
        return URL(string: collectionImage)
    }
}

File: android/VelaWallet/app/src/main/java/app/getvela/wallet/service/WalletApiService.kt
package app.getvela.wallet.service

import android.util.Log
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.net.HttpURLConnection
import java.net.URL

private const val TAG = "WalletAPI"
private const val BASE_URL = "https://getvela.app/api"

private val json = Json { ignoreUnknownKeys = true }

class WalletApiService {

    suspend fun fetchTokens(address: String): List<ApiToken> {
        val url = "$BASE_URL/wallet?address=$address"
        val data = httpGet(url) ?: return emptyList()
        return try {
            val response = json.decodeFromString<WalletResponse>(data)
            val filtered = response.tokens.filter { !it.spam }
            Log.d(TAG, "/wallet: ${filtered.size} tokens, total $${String.format("%.2f", filtered.sumOf { it.usdValue })}")
            filtered
        } catch (e: Exception) {
            Log.e(TAG, "/wallet parse failed: ${e.message}")
            emptyList()
        }
    }

    suspend fun fetchNFTs(address: String): List<ApiNft> {
        val url = "$BASE_URL/nft?address=$address"
        val data = httpGet(url) ?: return emptyList()
        return try {
            val response = json.decodeFromString<NftResponse>(data)
            Log.d(TAG, "/nft: ${response.nfts.size} NFTs")
            response.nfts
        } catch (e: Exception) {
            Log.e(TAG, "/nft parse failed: ${e.message}")
            emptyList()
        }
    }

    suspend fun fetchExchangeRate(currency: String = "CNY"): Double {
        val url = "$BASE_URL/exchange-rate?currency=$currency"
        val data = httpGet(url) ?: return 0.0
        return try {
            json.decodeFromString<ExchangeRateResponse>(data).rate
        } catch (e: Exception) {
            0.0
        }
    }

    private fun httpGet(urlString: String): String? {
        return try {
            val connection = URL(urlString).openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.connectTimeout = 15_000
            connection.readTimeout = 15_000
            connection.useCaches = false
            if (connection.responseCode == 200) {
                connection.inputStream.bufferedReader().readText()
            } else {
                Log.e(TAG, "HTTP ${connection.responseCode} for $urlString")
                null
            }
        } catch (e: Exception) {
            Log.e(TAG, "Network error: ${e.message}")
            null
        }
    }
}

// MARK: - Response Types

@Serializable
data class WalletResponse(val tokens: List<ApiToken>)

@Serializable
data class NftResponse(val nfts: List<ApiNft>)

@Serializable
data class ExchangeRateResponse(val currency: String, val rate: Double)

// MARK: - API Token Model (matches iOS APIToken)

@Serializable
data class ApiToken(
    val network: String,
    val chainName: String,
    val symbol: String,
    val balance: String,
    val decimals: Int,
    val logo: String? = null,
    val name: String,
    val tokenAddress: String? = null,
    val priceUsd: Double? = null,
    val spam: Boolean = false,
) {
    val id: String get() = "${network}_${tokenAddress ?: "native"}_$symbol"
    val isNative: Boolean get() = tokenAddress == null
    val balanceDouble: Double get() = balance.toDoubleOrNull() ?: 0.0
    val usdValue: Double get() = (priceUsd ?: 0.0) * balanceDouble

    val chainId: Int
        get() = when (network) {
            "eth-mainnet" -> 1
            "arb-mainnet" -> 42161
            "base-mainnet" -> 8453
            "opt-mainnet" -> 10
            "matic-mainnet" -> 137
            "bnb-mainnet" -> 56
            "avax-mainnet" -> 43114
            else -> 1
        }

    val logoUrl: String?
        get() {
            if (!logo.isNullOrEmpty()) return logo
            if (isNative) return "https://ethereum-data.awesometools.dev/chainlogos/eip155-$chainId.png"
            tokenAddress?.let { addr ->
                return "https://ethereum-data.awesometools.dev/assets/eip155-$chainId/$addr/logo.png"
            }
            return null
        }
}

// MARK: - NFT Model (matches iOS APINFT)

@Serializable
data class ApiNft(
    val network: String,
    val chainName: String,
    val contractAddress: String,
    val tokenId: String,
    val name: String? = null,
    val description: String? = null,
    val image: String? = null,
    val tokenType: String,
    val collectionName: String? = null,
    val collectionImage: String? = null,
) {
    val id: String get() = "${network}_${contractAddress}_$tokenId"
    val displayName: String get() = name ?: "${collectionName ?: "NFT"} #$tokenId"

    val imageUrl: String?
        get() {
            if (image.isNullOrEmpty()) return null
            if (image.startsWith("ipfs://")) return "https://ipfs.io/ipfs/${image.removePrefix("ipfs://")}"
            return image
        }
}
